### DISCUSSION TITLE
Piping produce, redux, devtools, logger and create

### DISCUSSION THREAD

**Discussion by GZLiew** [pmndrs/zustand#278] [Q&A]
Hi does  anyone know if this is possible ?

`const createStore = pipe(redux, immerProduce, devtools, logger, create`

Im getting a type error for immer

![image](https://user-images.githubusercontent.com/43195293/103971388-5e640900-51a5-11eb-82f5-df40a0639598.png)

heres my immer function 

![image](https://user-images.githubusercontent.com/43195293/103971564-eea24e00-51a5-11eb-8d7b-e57b2c23cf4f.png)

**Comment by harshagarwalnyu** [comment]
The type error when piping `immer` is a known friction point from zustand v3. The `pipe` utility's TypeScript inference breaks because `immer`'s overloaded types don't compose cleanly with the generic middleware signature.

**Modern fix (zustand v4+) — nest middleware directly:**

Zustand v4 replaced `pipe` with direct nesting:

```ts
import { create } from 'zustand';
import { immer } from 'zustand/middleware/immer';
import { devtools, redux } from 'zustand/middleware';

const useStore = create()(
  devtools(
    immer(
      redux(reducer, initialState)
    )
  )
);
```

**Why nesting works but `pipe` breaks:** Each middleware is a higher-order function wrapping the store creator. TypeScript propagates types through direct nesting. `pipe` requires all middleware to share a compatible generic signature — `immer`'s signature differs (it transforms the `set` function type).

**For the exact pipe pattern (zustand v3 workaround):**

```ts
const createStore = pipe(
  redux,
  immer as any,  // type-cast: immer's overload conflicts with pipe's inference
  devtools,
  logger,
  create
);
```

**Recommended for new code — nesting with explicit type:**

```ts
interface State {
  count: number;
  increment: () => void;
}

const useStore = create<State>()(
  devtools(
    immer((set) => ({
      count: 0,
      increment: () => set((state) => { state.count++ }),
    }))
  )
);
```

Reading order for nesting: outermost middleware wraps last (runs first on external calls). So `devtools(immer(...))` means devtools intercepts first, then immer transforms `set` before it reaches your store logic.
